src/hotspot/share/gc/shenandoah/shenandoahScanRemembered.inline.hpp

```diff
@@ -293,6 +376,8 @@ ShenandoahCardCluster<RememberedSet>::block_start(const size_t card_index) const
     assert(oopDesc::is_oop(cast_to_oop(left)), "Should be an object");
     return left;
   }
+  // If memory has not yet been allocated within card card_index, the above test will fail, but it still may be the
```

Comment:
One assumption here is that we ask this only of a dirty card. (May be we can assert that as a pre-condition?) Is it ever the case that we have a dirty card but in which no object has been allocated yet? When does that happen?

Reply:
That's a good question.  I'm not sure we're to the bottom of this yet.  It may be that this was a problem of my own creation, resulting from queries introduced by my "enhanced debugging instrumentation".  But I don't think that's the entire problem, because we have been seeing assert failures at the bottom of block_start() in our standard regression testing.

Reply:
Answers to your questions:
1. There can be dirty cards even if no object is allocated within a card's memory span.  This happens if a large reference array is allocated before the card's start address and ends after the card's start address, and an array entry that is within this card's span is overwritten with a pointer to young.
2. I added the assert you suggested, to confirm that the card being probed by block_start() is dirty.  This assert failed many times during tier0 jtreg testing.  I'll tabulate the failures and stack backtraces as soon as the test run finishes.

Reply:
There were 15 failures of this assert, all with a similar stack backtrace:

`
#  assert(_rs->is_card_dirty(card_index)) failed: Only ask block_start() if card is dirty

V  [libjvm.so+0x174ac1d]  ShenandoahCardCluster<ShenandoahDirectCardMarkRememberedSet>::block_start(unsigned long) const+0x57d  (shenandoahScanRemembered.inline.hpp:296)
V  [libjvm.so+0x174b230]  void ShenandoahScanRemembered<ShenandoahDirectCardMarkRememberedSet>::process_clusters<ShenandoahConcUpdateRefsClosure>(unsigned long, unsigned long, HeapWor\
dImpl**, ShenandoahConcUpdateRefsClosure*, bool, unsigned int)+0x570  (shenandoahScanRemembered.inline.hpp:690)
V  [libjvm.so+0x174d335]  void ShenandoahUpdateHeapRefsTask<true>::do_work<ShenandoahConcUpdateRefsClosure>(unsigned int)+0xf25  (shenandoahScanRemembered.inline.hpp:889)
V  [libjvm.so+0x174d663]  ShenandoahUpdateHeapRefsTask<true>::work(unsigned int)+0x73  (shenandoahHeap.cpp:2514)
V  [libjvm.so+0x1af77f0]  WorkerThread::run()+0x80  (workerThread.cpp:70)
V  [libjvm.so+0x19b5cea]  Thread::call_run()+0xba  (thread.cpp:221)
V  [libjvm.so+0x150306a]  thread_native_entry(Thread*)+0x12a  (os_linux.cpp:853)
`

Reply:
The assert failures were apparently due to confusion over is_card_dirty() vs is_write_card_dirty().  Resolving this issue, no more assertion failures in tier0 testing.  Going to test this version of code more broadly.